ert_shared/ensemble_evaluator/queue_adaptor.py

```diff
@@ -0,0 +1,121 @@
+        # events will be pushed.
+        self._publisher_thread = threading.Thread(
+            target=self._publisher, args=(asyncio.get_event_loop(),)
+        )
+        self._publisher_thread.start()
+
+    def _publisher(self, loop):
+        loop.run_until_complete(
+            self._publish_from_queue(self._changes_queue, self._ws_url)
+        )
+
+    @staticmethod
+    async def _publish_from_queue(queue, url):
+        while True:
+            async with websockets.connect(url) as websocket:
+                changes = await queue.get()
+                event = json.dumps(changes, cls=_EventDecoder)
+                await websocket.send(event)
+                if changes is None:
+                    return
+
+    def _initialize_state(self):
+        self._qindex_to_iens = {
+            q_index: q_node.callback_arguments[0].iens
+            for q_index, q_node in enumerate(self._queue.job_list)
+        }
+        self._state = [q_node.status.value for q_node in self._queue.job_list]
+
+    def _changes_after_transition(self):
+        """Return changes from transitioning to new state"""
+        new_state = [None] * len(self._state)
+        for q_index, job in enumerate(self.queue.job_list):
+            new_state[q_index] = job.status.value
+
+        changes = {}
+
+        diff = list(map(status_equal, zip(self._state, new_state)))
+        if len(diff) > 0:
+            for q_index, equal in enumerate(diff):
+                if not equal:
+                    st = str(JobStatusType(new_state[q_index]))
+                    changes[self._qindex_to_iens[q_index]] = create_realization(
+                        st, None
+                    )
+        self._state = new_state
+        return changes
+
+    def _snapshot(self):
+        snapshot = {}
+        for q_index, state_val in enumerate(self._state):
+            st = str(JobStatusType(state_val))
+            snapshot[self._qindex_to_iens[q_index]] = create_realization(st, None)
+        return snapshot
+
+    def _publish_changes(self, changes):
+        """changes is a dict where each item is a iens to _Realization map.
+        changes can be None, indicating EOT."""
+        if changes is not None and len(changes) == 0:
+            return
+
+        asyncio.run_coroutine_threadsafe(
+            self._changes_queue.put(changes), asyncio.get_event_loop()
+        ).result()
+
+    def execute_queue(self):
+        while self.queue.is_active() and not self.queue.stopped:
+            self._launch_jobs()
+
+            time.sleep(1)
+
+            if self._queue_evaluators is not None:
+                for func in self._queue_evaluators:
+                    func()
+
+            self._publish_changes(self._changes_after_transition())
+
+        if self.queue.stopped:
+            self._stop_jobs()
+
+        self._publish_changes(self._snapshot())
+
+        self._assert_complete()
+
+        self._publish_changes(None)
```

Comment:
Do we need to publish a snapshot at the start also, or are we safe since `_initialize_state` is called in the init function and not after the Queue is started?